ddl/ddl_api.go

```diff
@@ -1424,23 +1425,35 @@ func buildTableInfo(
 			if err != nil {
 				return nil, err
 			}
-			if !config.GetGlobalConfig().AlterPrimaryKey {
-				singleIntPK := isSingleIntPK(constr, lastCol)
-				clusteredIdx := ctx.GetSessionVars().EnableClusteredIndex
-				if singleIntPK || clusteredIdx {
-					// Primary key cannot be invisible.
-					if constr.Option != nil && constr.Option.Visibility == ast.IndexVisibilityInvisible {
-						return nil, ErrPKIndexCantBeInvisible
-					}
-				}
-				if singleIntPK {
+			pkTp := model.PrimaryKeyTypeDefault
+			if constr.Option != nil {
+				pkTp = constr.Option.PrimaryKeyTp
+			}
+			switch pkTp {
+			case model.PrimaryKeyTypeNonClustered:
+				break
+			case model.PrimaryKeyTypeClustered:
+				if isSingleIntPK(constr, lastCol) {
 					tbInfo.PKIsHandle = true
-					// Avoid creating index for PK handle column.
-					continue
-				}
-				if clusteredIdx {
+				} else {
 					tbInfo.IsCommonHandle = true
 				}
+			case model.PrimaryKeyTypeDefault:
+				alterPKConf := config.GetGlobalConfig().AlterPrimaryKey
+				if isSingleIntPK(constr, lastCol) {
+					tbInfo.PKIsHandle = !alterPKConf
+				} else {
+					tbInfo.IsCommonHandle = !alterPKConf && ctx.GetSessionVars().EnableClusteredIndex
```

Comment:
Shall we return a warning if `model.PrimaryKeyTypeClustered && !ctx.GetSessionVars().EnableClusteredIndex`?

Reply:
No, the variable has lower priority.